File: tests/node/test_fal_smoke.mjs
/**
 * fal.ai compute smoke tests.
 *
 * Gated on the FAL_API_KEY environment variable.
 * Tests LLM completion, modality auto-routing, streaming, embeddings,
 * 3D generation, image utilities, and timing/cost metadata.
 */

// Invocation examples:
//   default (LLM only):        node --test tests/node/test_fal_smoke.mjs
//   + compute (image/TTS/...): BLAZEN_TEST_FAL_COMPUTE=1 node --test tests/node/test_fal_smoke.mjs
//   + video:                   BLAZEN_TEST_FAL_VIDEO=1 node --test tests/node/test_fal_smoke.mjs

import test from "ava";

import {
  CompletionModel,
  ChatMessage,
  FalProvider,
} from "../../crates/blazen-node/index.js";
import { falOrSkip, expectFalRoutingError } from "./_smoke_helpers.mjs";

const FAL_API_KEY = process.env.FAL_API_KEY;
const FAL_COMPUTE = process.env.BLAZEN_TEST_FAL_COMPUTE === "1";
const FAL_VIDEO = process.env.BLAZEN_TEST_FAL_VIDEO === "1";

const IMAGE_URL =
  "https://upload.wikimedia.org/wikipedia/commons/thumb/4/47/PNG_transparency_demonstration_1.png/280px-PNG_transparency_demonstration_1.png";
// 3D generation deliberately uses an unfetchable URL (`example.invalid` is
// reserved by RFC 2606) so fal fast-fails with file_download_error rather
// than running full triposr 3D gen, which regularly exceeds the 300s test
// budget on fal's queue. The test only cares about routing verification.
const UNFETCHABLE_IMAGE_URL = "https://example.invalid/triposr_input.jpg";
const AUDIO_URL = "https://upload.wikimedia.org/wikipedia/commons/c/c8/Example.ogg";
const VIDEO_URL =
  "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";

const TLlm = FAL_API_KEY ? test : test.skip;
const TRoute = FAL_API_KEY ? test : test.skip;
const TStream = FAL_API_KEY ? test : test.skip;
const TCompute = FAL_API_KEY && FAL_COMPUTE ? test : test.skip;

TLlm("fal.ai LLM smoke tests · completes a basic prompt via OpenAiChat default", async (t) => {
  const model = CompletionModel.fal({ apiKey: FAL_API_KEY });
  const response = await model.complete([
    ChatMessage.user("What is 2+2? Reply with just the number."),
  ]);

  t.truthy(response.content, "expected content in response");
  t.truthy(response.content.includes("4"), `expected '4' in: ${response.content}`);
  t.truthy(response.model, "expected model in response");
});

TLlm("fal.ai LLM smoke tests · completes a basic prompt with enterprise=true", async (t) => {
  const model = CompletionModel.fal({ apiKey: FAL_API_KEY, enterprise: true });
  const response = await model.complete([
    ChatMessage.user("What is 3+3? Reply with just the number."),
  ]);

  t.truthy(response.content, "expected content in response");
  t.truthy(response.content.includes("6"), `expected '6' in: ${response.content}`);
});

TLlm("fal.ai LLM smoke tests · completes a basic prompt via OpenAiResponses endpoint", async (t) => {
  const model = CompletionModel.fal({
    apiKey: FAL_API_KEY,
    endpoint: "open_ai_responses",
  });
  const response = await model.complete([
    ChatMessage.user("What is 5+5? Reply with just the number."),
  ]);

  t.truthy(response.content, "expected content in response");
  t.truthy(response.content.includes("10"), `expected '10' in: ${response.content}`);
});

TLlm("fal.ai LLM smoke tests · returns timing metadata", async (t) => {
  const model = CompletionModel.fal({ apiKey: FAL_API_KEY });
  const response = await model.complete([
    ChatMessage.user("Say hello."),
  ]);

  t.truthy(response.content, "expected content");
  // Timing should be populated for fal.ai queue mode
  if (response.timing) {
    t.truthy(
      typeof response.timing.totalMs === "number" || response.timing.totalMs === null,
      "timing.totalMs should be a number or null"
    );
  } else {
    t.pass("no timing metadata returned");
  }
});

TLlm("fal.ai LLM smoke tests · passes temperature and max_tokens", async (t) => {
  const model = CompletionModel.fal({ apiKey: FAL_API_KEY });
  const response = await model.completeWithOptions(
    [ChatMessage.user("Write a one-word greeting.")],
    { temperature: 0.1, maxTokens: 10 }
  );

  t.truthy(response.content, "expected content");
  // With max_tokens=10, response should be short
  t.truthy(response.content.length < 200, "response should be short with maxTokens=10");
});

TLlm("fal.ai LLM smoke tests · exposes new typed CompletionResponse fields (reasoning/citations/artifacts)", async (t) => {
  const model = CompletionModel.fal({ apiKey: FAL_API_KEY });
  const response = await model.complete([
    ChatMessage.user("Say hi."),
  ]);

  t.truthy(response.content, "expected content");
  // Reasoning is Option<> on the Rust side; napi serializes None as undefined,
  // which means the property may not be enumerable via `in`. Just check that
  // accessing it doesn't throw and that citations/artifacts are arrays.
  t.truthy(
    response.reasoning === undefined || typeof response.reasoning === "object",
    "expected 'reasoning' to be undefined or object"
  );
  t.truthy(
    Array.isArray(response.citations),
    "expected 'citations' field on CompletionResponse"
  );
  t.truthy(
    Array.isArray(response.artifacts),
    "expected 'artifacts' field on CompletionResponse"
  );
});

TRoute("fal.ai modality auto-routing tests · auto-routes to vision when AnyLlm and image part present", async (t) => {
  const model = CompletionModel.fal({
    apiKey: FAL_API_KEY,
    endpoint: "any_llm",
    autoRouteModality: true,
  });
  const response = await model.complete([
    ChatMessage.userImageUrl("Describe this image in one short sentence.", IMAGE_URL, "image/png"),
  ]);

  t.truthy(response.content, "expected content from vision auto-route");
  t.truthy(response.content.length > 0, "expected non-empty vision response");
});

TRoute("fal.ai modality auto-routing tests · auto-routes to audio when OpenRouter and audio part present", async (t) => {
  const model = CompletionModel.fal({
    apiKey: FAL_API_KEY,
    endpoint: "open_router",
    autoRouteModality: true,
  });
  const msg = ChatMessage.userAudio("Transcribe or describe this audio briefly.", AUDIO_URL);
  const response = await expectFalRoutingError(
    model.complete([msg]),
    ["Failed to download audio", "audio_url", "file_download_error"],
  );
  if (response !== undefined) {
    t.truthy(response.content, "expected content from audio auto-route");
    t.truthy(response.content.length > 0, "expected non-empty audio response");
  } else {
    t.pass("expected fal routing error observed");
  }
});

TRoute("fal.ai modality auto-routing tests · auto-routes to video when OpenRouter and video part present", async (t) => {
  const model = CompletionModel.fal({
    apiKey: FAL_API_KEY,
    endpoint: "open_router",
    autoRouteModality: true,
  });
  const msg = ChatMessage.userVideo("Describe this video in one short sentence.", VIDEO_URL);
  const response = await expectFalRoutingError(
    model.complete([msg]),
    ["Failed to download video", "video_url", "file_download_error"],
  );
  if (response !== undefined) {
    t.truthy(response.content, "expected content from video auto-route");
    t.truthy(response.content.length > 0, "expected non-empty video response");
  } else {
    t.pass("expected fal routing error observed");
  }
});

TStream("fal.ai streaming + embeddings + utilities · streaming yields multiple chunks", async (t) => {
  const model = CompletionModel.fal({ apiKey: FAL_API_KEY });
  const chunks = [];

  await model.stream(
    [ChatMessage.user("Count from 1 to 5, one number per line.")],
    (chunk) => {
      chunks.push(chunk);
    },
  );

  t.truthy(chunks.length > 0, `expected at least one streamed chunk, got ${chunks.length}`);
});

TStream("fal.ai streaming + embeddings + utilities · FalProvider.embeddingModel().embed produces vectors", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const embedder = provider.embeddingModel();
  const vectors = await embedder.embed(["hi", "hello"]);

  t.truthy(Array.isArray(vectors), "expected an array of vectors");
  t.is(vectors.length, 2, "expected 2 vectors");
  t.is(vectors[0].length, 1536, "expected vector dim 1536");
  t.is(vectors[1].length, 1536, "expected vector dim 1536");
});

TCompute("fal.ai streaming + embeddings + utilities · generates a 3D model", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await expectFalRoutingError(
    provider.generate3d({ imageUrl: UNFETCHABLE_IMAGE_URL }),
    ["Failed to download", "image_url", "file_download_error"],
  );
  if (result !== undefined) {
    t.truthy(result, "expected a 3D generation result");
  } else {
    t.pass("expected fal routing error observed");
  }
});

TCompute("fal.ai streaming + embeddings + utilities · removes background from an image", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await expectFalRoutingError(
    provider.removeBackground({ imageUrl: IMAGE_URL }),
    ["Failed to download", "image_url", "file_download_error"],
  );
  if (result !== undefined) {
    t.truthy(result, "expected a background-removal result");
  } else {
    t.pass("expected fal routing error observed");
  }
});

TCompute("fal.ai compute smoke tests · generates an image with FLUX", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await falOrSkip(t, provider.generateImage({ prompt: "a simple red circle on white" }));
  if (result === undefined) return;
  t.truthy(result.images, "expected images");
  t.truthy(result.images.length > 0, "expected at least one image");
});

TCompute("fal.ai compute smoke tests · synthesizes speech from text", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await falOrSkip(t, provider.textToSpeech({ text: "Hello world." }));
  if (result === undefined) return;
  t.truthy(result, "expected a result");
  t.truthy(
    result.audio_url || result.audio || result.url,
    `expected audio data in result, got keys: ${Object.keys(result).join(", ")}`
  );
});

TCompute("fal.ai compute smoke tests · generates music from a prompt", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await falOrSkip(t, provider.generateMusic({
    prompt: "happy upbeat jingle",
    durationSeconds: 5,
  }));
  if (result === undefined) return;
  t.truthy(result, "expected a result");
  t.truthy(
    result.audio_url || result.audio || result.url || result.audio_file,
    `expected audio data in result, got keys: ${Object.keys(result).join(", ")}`
  );
});

TCompute("fal.ai compute smoke tests · transcribes audio from a URL", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await falOrSkip(t, provider.transcribe({
    audioUrl: "https://cdn.openai.com/API/docs/audio/alloy.wav",
  }));
  if (result === undefined) return;
  t.truthy(result, "expected a result");
  t.truthy(
    result.text || result.transcript || result.chunks,
    `expected text content in result, got keys: ${Object.keys(result).join(", ")}`
  );
});

const TVideo = FAL_API_KEY && FAL_COMPUTE && FAL_VIDEO ? test : test.skip;

TVideo("fal.ai compute smoke tests · generates a video from a text prompt", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await falOrSkip(t, provider.textToVideo({ prompt: "a cat walking slowly" }));
  if (result === undefined) return;

  t.truthy(result, "expected a result");
  t.truthy(
    result.videos || result.video_url || result.video || result.url,
    `expected video data in result, got keys: ${Object.keys(result).join(", ")}`
  );
  if (result.videos) {
    t.truthy(result.videos.length > 0, "expected at least one video");
  }
});

TCompute("fal.ai compute smoke tests · runs a model synchronously via run()", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });
  const result = await falOrSkip(t, provider.run({
    model: "fal-ai/flux/schnell",
    input: { prompt: "blue sky with white clouds", imageSize: "square_hd" },
  }));
  if (result === undefined) return;
  t.truthy(result, "expected a non-null result from run()");
});

TCompute("fal.ai compute smoke tests · submits a job and gets a valid job handle", async (t) => {
  const provider = FalProvider.create({ apiKey: FAL_API_KEY });

  const job = await falOrSkip(t, provider.submit({
    model: "fal-ai/flux/schnell",
    input: { prompt: "green forest with sunlight" },
  }));
  if (job === undefined) return;

  t.truthy(job, "expected a job handle from submit()");

  const jobId = job.id || job.requestId || job.request_id;
  t.truthy(jobId, `expected job id, got keys: ${Object.keys(job).join(", ")}`);
  t.truthy(typeof jobId === "string" && jobId.length > 0, "job id should be a non-empty string");
});

File: tests/python/test_fal_smoke.py
"""fal.ai compute smoke tests.

Gated on the FAL_API_KEY environment variable.
Tests LLM completion, image generation, TTS, transcription, and timing metadata.
"""

import os

import pytest

from blazen import (
    BackgroundRemovalRequest,
    ChatMessage,
    CompletionModel,
    CompletionOptions,
    ComputeRequest,
    FalLlmEndpointKind,
    FalOptions,
    FalProvider,
    ImageRequest,
    MusicRequest,
    RetryConfig,
    SpeechRequest,
    ThreeDRequest,
    TranscriptionRequest,
    VideoRequest,
)

FAL_API_KEY = os.environ.get("FAL_API_KEY")

skip_without_key = pytest.mark.skipif(
    not FAL_API_KEY,
    reason="FAL_API_KEY not set",
)


async def _fal_or_skip(coro):
    """Await a fal compute coroutine; skip on transient infra errors."""
    try:
        return await coro
    except Exception as e:
        err = str(e)
        if any(s in err for s in ("502", "503", "504", "Bad Gateway", "service_unavailable")):
            pytest.skip(f"fal.ai infra error: {err}")
        raise


# ---------------------------------------------------------------------------
# LLM completion via fal-ai/any-llm
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_completion_openai_chat():
    """Default endpoint (OpenAiChat) with temperature, max_tokens, and response field checks."""
    model = CompletionModel.fal(options=FalOptions(api_key=FAL_API_KEY)).with_retry(RetryConfig(max_retries=2))
    response = await model.complete(
        [ChatMessage.user("What is 2+2? Reply with just the number.")],
        CompletionOptions(temperature=0.1, max_tokens=10),
    )
    assert response.content is not None
    assert "4" in response.content
    assert response.model is not None
    assert hasattr(response, "tool_calls")
    assert hasattr(response, "reasoning")
    assert hasattr(response, "citations")
    assert hasattr(response, "artifacts")


@skip_without_key
@pytest.mark.asyncio
async def test_fal_basic_completion_enterprise():
    """Enterprise mode promotes OpenAiChat -> AnyLlm{enterprise:true}."""
    model = CompletionModel.fal(options=FalOptions(api_key=FAL_API_KEY, enterprise=True)).with_retry(RetryConfig(max_retries=2))
    response = await model.complete(
        [ChatMessage.user("Say hello.")],
        CompletionOptions(max_tokens=10),
    )
    assert response.content is not None


@skip_without_key
@pytest.mark.asyncio
async def test_fal_responses_api_endpoint():
    """OpenAiResponses endpoint targets openrouter/router/openai/v1/responses."""
    model = CompletionModel.fal(
        options=FalOptions(api_key=FAL_API_KEY, endpoint=FalLlmEndpointKind.OpenAiResponses),
    ).with_retry(RetryConfig(max_retries=2))
    response = await model.complete(
        [ChatMessage.user("Say hi.")],
        CompletionOptions(max_tokens=10),
    )
    assert response.content is not None


# ---------------------------------------------------------------------------
# Modality auto-routing (AnyLlm endpoint)
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_vision_auto_routes_when_anyllm_and_image_present():
    """When AnyLlm endpoint is configured and a message has an image part,
    the request should auto-route to fal-ai/any-llm/vision."""
    model = CompletionModel.fal(
        options=FalOptions(api_key=FAL_API_KEY, endpoint=FalLlmEndpointKind.AnyLlm),
    ).with_retry(RetryConfig(max_retries=2))
    msg = ChatMessage.user_image_url(
        text="What is in this image? One word.",
        url="https://upload.wikimedia.org/wikipedia/commons/thumb/4/47/PNG_transparency_demonstration_1.png/280px-PNG_transparency_demonstration_1.png",
        media_type="image/png",
    )
    response = await model.complete([msg], CompletionOptions(max_tokens=20))
    assert response.content is not None


@skip_without_key
@pytest.mark.asyncio
async def test_fal_audio_auto_routes_when_openrouter_and_audio_present():
    """Audio LLM auto-routes from OpenRouter to openrouter/router/audio.

    fal's workers cannot fetch files from arbitrary public CDNs in test
    environments, so this test asserts that EITHER the call succeeds OR
    the error is specifically a fal file-download failure (which proves
    routing landed at the audio endpoint correctly).
    """
    model = CompletionModel.fal(
        options=FalOptions(api_key=FAL_API_KEY, endpoint=FalLlmEndpointKind.OpenRouter),
    )
    msg = ChatMessage.user_audio(
        text="What does this clip say?",
        url="https://upload.wikimedia.org/wikipedia/commons/c/c8/Example.ogg",
    )
    try:
        response = await model.complete([msg], CompletionOptions(max_tokens=30))
        assert response.content is not None
    except Exception as e:
        err = str(e)
        # Routing succeeded if fal accepted the request and tried to fetch the audio.
        assert any(
            marker in err
            for marker in (
                "Failed to download audio",
                "audio_url",
                "file_download_error",
            )
        ), f"unexpected error (routing may have failed): {err}"


@skip_without_key
@pytest.mark.asyncio
async def test_fal_video_auto_routes_when_openrouter_and_video_present():
    """Video LLM auto-routes from OpenRouter to openrouter/router/video.

    fal's workers cannot fetch files from arbitrary public CDNs in test
    environments, so this test asserts that EITHER the call succeeds OR
    the error is specifically a fal file-download failure (which proves
    routing landed at the video endpoint correctly).
    """
    model = CompletionModel.fal(
        options=FalOptions(api_key=FAL_API_KEY, endpoint=FalLlmEndpointKind.OpenRouter),
    )
    msg = ChatMessage.user_video(
        text="What is happening in this video? One sentence.",
        url="https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4",
    )
    try:
        response = await model.complete([msg], CompletionOptions(max_tokens=40))
        assert response.content is not None
    except Exception as e:
        err = str(e)
        # Routing succeeded if fal accepted the request and tried to fetch the video.
        assert any(
            marker in err
            for marker in (
                "Failed to download video",
                "video_url",
                "file_download_error",
            )
        ), f"unexpected error (routing may have failed): {err}"


# ---------------------------------------------------------------------------
# Streaming, embeddings, 3D, background removal
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_streaming_yields_multiple_chunks():
    model = CompletionModel.fal(options=FalOptions(api_key=FAL_API_KEY)).with_retry(RetryConfig(max_retries=2))
    chunks = []
    async for chunk in model.stream(
        [ChatMessage.user("Count from 1 to 5, one number per line.")],
        options=CompletionOptions(max_tokens=50),
    ):
        chunks.append(chunk)
    assert len(chunks) > 1, f"expected multiple chunks, got {len(chunks)}"


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_embeddings():
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    em = provider.embedding_model()
    response = await _fal_or_skip(em.embed(["hello", "world"]))
    assert len(response["embeddings"]) == 2
    assert len(response["embeddings"][0]) == 1536


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_3d_generation():
    """3D generation routes to the correct fal endpoint.

    fal's workers cannot fetch files from arbitrary public CDNs in test
    environments, so this test asserts that EITHER the call succeeds OR
    the error is specifically a fal file-download failure (which proves
    routing landed at the 3D endpoint correctly).
    """
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    request = ThreeDRequest(
        prompt="a wooden chair",
        image_url="https://storage.googleapis.com/falserverless/example_inputs/triposr_input.jpg",
    )
    try:
        result = await _fal_or_skip(provider.generate_3d(request))
        assert result is not None
    except Exception as e:
        err = str(e)
        # Routing succeeded if fal accepted the request and tried to fetch the image.
        assert any(
            marker in err
            for marker in (
                "file_download_error",
                "image_url",
            )
        ), f"unexpected error (routing may have failed): {err}"


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_background_removal():
    """Background removal routes to the correct fal endpoint.

    fal's workers cannot fetch files from arbitrary public CDNs in test
    environments, so this test asserts that EITHER the call succeeds OR
    the error is specifically a fal file-download failure (which proves
    routing landed at the background-removal endpoint correctly).
    """
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    try:
        result = await _fal_or_skip(provider.remove_background(
            BackgroundRemovalRequest(image_url="https://storage.googleapis.com/falserverless/example_inputs/birefnet_input.jpeg"),
        ))
        assert result is not None
    except Exception as e:
        err = str(e)
        # Routing succeeded if fal accepted the request and tried to fetch the image.
        assert any(
            marker in err
            for marker in (
                "file_download_error",
                "image_url",
            )
        ), f"unexpected error (routing may have failed): {err}"


# ---------------------------------------------------------------------------
# FalProvider compute tests
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_image_generation():
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    result = await _fal_or_skip(provider.generate_image(
        ImageRequest(prompt="a simple red circle on white background"),
    ))
    assert result.images is not None
    assert len(result.images) > 0


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(90)
async def test_fal_text_to_speech():
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    result = await _fal_or_skip(provider.text_to_speech(
        SpeechRequest(text="Hello world, this is a test."),
    ))
    assert result.audio is not None
    assert len(result.audio) > 0


# ---------------------------------------------------------------------------
# Compute tests (music, transcription, video, raw run, job lifecycle)
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_generate_music():
    """Generate music and verify the response contains audio data."""
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    result = await _fal_or_skip(provider.generate_music(
        MusicRequest(prompt="happy upbeat jingle", duration_seconds=5.0),
    ))
    assert result.audio is not None
    assert len(result.audio) > 0


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_transcribe():
    """Transcribe a short audio clip and verify text is returned."""
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    result = await _fal_or_skip(provider.transcribe(
        TranscriptionRequest(audio_url="https://cdn.openai.com/API/docs/audio/alloy.wav"),
    ))
    assert result.text is not None
    assert len(result.text) > 0


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(600)
async def test_fal_text_to_video():
    """Generate a video from text (slow, ~30-60s)."""
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    try:
        result = await _fal_or_skip(provider.text_to_video(
            VideoRequest(prompt="a cat walking"),
        ))
    except Exception as e:
        err = str(e).lower()
        if any(
            marker in err
            for marker in (
                "downstream_service_unavailable",
                "service unavailable",
                "service_unavailable",
            )
        ):
            pytest.skip(f"fal.ai video service transiently unavailable: {e}")
        raise
    assert result.videos is not None
    assert len(result.videos) > 0


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_raw_compute_run():
    """Run a raw compute job via the generic run() method."""
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    result = await _fal_or_skip(provider.run(
        ComputeRequest(
            model="fal-ai/flux/schnell",
            input={"prompt": "blue sky", "image_size": "square_hd"},
        ),
    ))
    assert result is not None
    assert result.output is not None


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(300)
async def test_fal_job_submit():
    """Submit a job and verify a job handle is returned."""
    provider = FalProvider(options=FalOptions(api_key=FAL_API_KEY))
    job = await _fal_or_skip(provider.submit(
        ComputeRequest(
            model="fal-ai/flux/schnell",
            input={"prompt": "green forest"},
        ),
    ))
    assert job.id is not None
    assert isinstance(job.id, str)
    assert len(job.id) > 0
